newrelic/api/opentelemetry.py

```diff
@@ -0,0 +1,476 @@
+        raise NotImplementedError("Not implemented yet")
+
+    def record_exception(self, exception, attributes=None, timestamp=None, escaped=False):
+        error_args = sys.exc_info() if not exception else (type(exception), exception, exception.__traceback__)
+
+        if not hasattr(self, "nr_trace"):
+            notice_error(error_args, attributes=attributes)
+        else:
+            self.nr_trace.notice_error(error_args, attributes=attributes)
+
+    def end(self, end_time=None, *args, **kwargs):
+        # We will ignore the end_time parameter and use NR's end_time
+
+        # Check to see if New Relic trace ever existed or,
+        # if it does, that trace has already ended
+        nr_trace = hasattr(self, "nr_trace")
+        if not nr_trace or (nr_trace and getattr(nr_trace, "end_time", None)):
+            return
+
+        # Add attributes as Trace parameters
+        self._set_attributes_in_nr(self.attributes)
+
+        # For each kind of NR Trace, we will need to add
+        # specific attributes since they were likely not
+        # available at the time of the trace's creation.
+        if self.instrumenting_module in ("Redis", "Mongodb"):
+            self.nr_trace.host = self.attributes.get("net.peer.name", self.attributes.get("server.address"))
+            self.nr_trace.port_path_or_id = self.attributes.get("net.peer.port", self.attributes.get("server.port"))
+            self.nr_trace.database_name = self.attributes.get("db.name")
+            self.nr_trace.product = self.attributes.get("db.system")
+        elif self.instrumenting_module == "Dynamodb":
+            self.nr_trace.database_name = self.attributes.get("db.name")
+            self.nr_trace.product = self.attributes.get("db.system")
+            self.nr_trace.port_path_or_id = self.attributes.get("net.peer.port")
+            self.nr_trace.host = self.attributes.get("dynamodb.{region}.amazonaws.com")
+
+        # Set SpanKind attribute
+        self._set_attributes_in_nr({"span.kind": self.kind})
+
+        self.nr_trace.__exit__(*sys.exc_info())
+
+
+class Tracer(otel_api_trace.Tracer):
+    def __init__(self, resource=None, instrumentation_library=None, *args, **kwargs):
+        self.resource = resource
+        self.instrumentation_library = instrumentation_library.split(".")[-1].capitalize()
+
+    def start_span(
+        self,
+        name,
+        context=None,  # Optional[Context]
+        kind=otel_api_trace.SpanKind.INTERNAL,
+        attributes=None,
+        links=None,
+        start_time=None,
+        record_exception=True,
+        set_status_on_exception=True,
+        *args,
+        **kwargs,
+    ):
+        # Retrieve parent span
+        parent_span_context = otel_api_trace.get_current_span(context).get_span_context()
+
+        if parent_span_context is None or not parent_span_context.is_valid:
+            parent_span_context = None
+
+        nr_trace_type = FunctionTrace
+        transaction = current_transaction()
```

Comment:
Wondering if it would be wise to use this transaction to make sure we don't have one that already exists and is active before we create a new one?

Reply:
The agent creates a trace (in the form of a Sentinel) if a transaction already exists and will not create a new transaction, so I did not create separate logic for that part of it